README.md

```diff
@@ -1,279 +1,106 @@
 
-`ember-deploy` is built around the idea of adapters for the bootstrap-index- and the assets-uploads. For `ember-deploy` to give you the deployment approach Luke talks about in his talk you will for example need to install the `ember-deploy-redis` and `ember-deploy-s3` adapters via npm.
+Plugins are npm packages that hook into the deployment pipeline and actually perform tasks.  An example of an ember-cli-deploy plugin might be a plugin that builds the project, uploads the index.html to redis or uploads assets to S3.
 
-Because `ember-deploy` is built with adapters in mind you can write your own adapters for your project specific use cases.
+In order for an npm package to be registered as an ember-cli-deploy plugin, it must be an [Ember CLI addon](http://www.ember-cli.com/#developing-addons-and-blueprints) and must specify the `ember-ci-deploy-plugin` keyword in it's `package.json` like so:
 
-Custom Adapters can be integrated via custom [Ember-CLI-Addons](http://www.ember-cli.com/#developing-addons-and-blueprints). For `ember-deploy` to integrate your custom adapter addons you have to define your addon-type as an `ember-deploy-addon`.
+```javascript
+// package.json
 
-You can have a look at [ember-deploy-s3](https://github.com/LevelbossMike/ember-deploy-s3) and [ember-deploy-redis](https://github.com/LevelbossMike/ember-deploy-redis) to get an idea how one would implement 'real-world'-adapters.
+"keywords": [
+  "ember-addon",
+  "ember-cli-deploy-plugin"
+]
+```
 
-For the sake of clarity here's a quick example of how you will structure your addon to add custom adapters.
+Finally the plugin must implement at least one of the above hooks.  The ember-cli-deploy pipeline will invoke any implemented hooks at the relevant point in the pipeline.
 
-```js
-//index.js in your custom addon
+Hooks must be self sufficient and not rely on any effects from other hooks. Each implemented hooks must also return a `Promise` which resolves or rejects depending on whether it ran successfully or not.
 
-var SuperAwesomeCustomIndexAdapter = require('./lib/index-adapter');
+The promise returned from a plugin hook should either:
 
-function EmberDeploySuperAwesome() {
-  this.name = 'ember-deploy-super-awesome';
-  this.type = 'ember-deploy-addon';
+- resolve with `undefined`;
+- reject with an `Error` instance if the error is unhandled or;
+- reject with `undefined` if it was handled.
 
-  // ember-deploy will merge ember-deploy-addon's adapters property
-  this.adapters = {
-    index: {
-      'super-awesome': SuperAwesomeCustomIndexAdapter
-    },
-    assets: {
-      // you can add multiple adapters in one ember-deploy-addon
-    },
+An ember-cli-deploy plugin should look something like this:
 
-    tagging: {
-      // you can add multiple adapters in one ember-deploy-addon
-    }
-  };
-}
+```javascript
+// plugins/lightening-fast-deployment.js
 
-module.exports = EmberDeploySuperAwesome;
-```
+var Plugin = require('ember-cli-deploy/models/plugin');
```

Comment:
There was a brief discussion about not having the plugins depend explicitly and the possibility of "...plugins implementing a willDeploy hook that essentially runs validation: e.g. is required  configuration present? that could include checking for a minimum version of ember-cli-deploy"

I did run into an issue yesterday where a plugin depended on ember-deploy and it stomped on ember-cli-deploy that I had installed. Not sure that would be an issue though with ember-cli-deploy in the future. Also, this potentially could break a plugin due to a version mismatch in the future or a weird cross dependency issue.

I mentioned peerDependencies but that could also be a can of worms...hence the thought of checking the version in a hook or something.